_posts/2023-04-30-writing-code-with-chatgpt-improve-it-with-kani.md

```diff
@@ -0,0 +1,359 @@
+
+``` rust
+// Original implementation.
+fn lowest_unset_bit_ori(x: u32) -> Option<u8> {
+    let mut mask = 1;
+    for i in 0..32 {
+        if x & mask == 0 {
+            return Some(i);
+        }
+        mask <<= 1;
+    }
+    None
+} 
+// Optimized implementation generated by ChatGPT.
+fn lowest_unset_bit_opt(x: u32) -> Option<u8> {
+    if x == std::u32::MAX{
+        return None;
+    }
+    let y = x | (x + 1);        // <------ line 18 
+    let z = y ^ (y + 1);
+    Some((z.trailing_zeros() % 32) as u8)
+} 
+
+#[kani::proof]
+fn check() {
+    let x: u32 = kani::any();
+    assert_eq!(lowest_unset_bit_ori(x), lowest_unset_bit_opt(x));   // <------ line 26
+}
+```
+
+Running Kani on the program with the command
+
+```
+> kani lowest_unset_bit.rs
+```
+
+gave us the verification result
+
+```
+...
+SUMMARY:
+ ** 2 of 70 failed
+Failed Checks: attempt to add with overflow
+ File: "...", line 18, in lowest_unset_bit_opt
+Failed Checks: assertion failed: lowest_unset_bit_ori(x) == lowest_unset_bit_opt(x)
+ File: "...", line 26, in check
+
+VERIFICATION:- FAILED
+Verification Time: 0.84979576s
+```
+
+The generated implementation is incorrect!
+There were two failed checks.
+The first one `Failed Checks: attempt to add with overflow` says that the addition `x + 1` may overflow.
+The other one says that equivalent check may fail---the two implementations are actually not equivalent.
+
+
+To better understand the bug, we asked Kani to produce a counterexample with the `--concrete-playback` option (discussed in [our previous post](https://model-checking.github.io/kani-verifier-blog/2022/09/22/internship-projects-2022-concrete-playback.html))
+
+```
+> cargo kani --enable-unstable --concrete-playback=print | grep "//" | awk '{print $2}'
+2147483647
+```
+
+The counterexample is when `x` equals `2147483647` (`i32::MAX`).
+
+
+The examples above illustrate that, although code generation is efficient and accessible, it's crucial to exercise caution and avoid blindly trusting the generated code, as it may contain errors.
+We believe there are two factors that contribute to this unreliability.
+Firstly, the training corpus for most LLMs is unverified code.
+Secondly, text prompts can be ambiguous, leading to the production of flawed code.
+Fortunately, program verifiers such as Kani can provide valuable information about generated code, such as verification results, and counterexamples.
+Armed with this information, developers can locate and fix bugs more effectively.
+But what about the potential for language models like ChatGPT to leverage such insights to improve code generation accuracy?
+It's an intriguing question that merits further exploration.
+
+
+
+
+## Code Generation with Verifiers in the Loop
+The last task we explore in this blog is: can we combine Kani and ChatGPT to generate correct and verified code?
```

Comment:
I find "Generating verified code" confusing. I'd rather phrase it as "generating code that passes verification".